api/build/traffic.go

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package build
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+
+	"github.com/go-vela/server/api/types"
+	"github.com/go-vela/server/constants"
+	"github.com/go-vela/server/database"
+	"github.com/go-vela/server/queue"
+	"github.com/go-vela/server/queue/models"
+	"github.com/go-vela/server/scm"
+)
+
+// TrafficBuild is a helper function that will determine whether to publish a build to the queue or place it
+// in pending approval status.
+func TrafficBuild(c *gin.Context, l *logrus.Entry, b *types.Build, r *types.Repo, item *models.Item) error {
+	// if the webhook was from a Pull event from a forked repository, verify it is allowed to run
+	if b.GetFork() {
+		l.Tracef("inside %s workflow for fork PR build %s/%d", r.GetApproveBuild(), r.GetFullName(), b.GetNumber())
+
+		switch r.GetApproveBuild() {
+		case constants.ApproveForkAlways:
+			err := gatekeepBuild(c, b, r)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		case constants.ApproveForkNoWrite:
+			// determine if build sender has write access to parent repo. If not, this call will result in an error
+			level, err := scm.FromContext(c).RepoAccess(c.Request.Context(), b.GetSender(), r.GetOwner().GetToken(), r.GetOrg(), r.GetName())
+			if err != nil || (level != "admin" && level != "write") {
+				err = gatekeepBuild(c, b, r)
+				if err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			l.Debugf("fork PR build %s/%d automatically running without approval. sender %s has %s access", r.GetFullName(), b.GetNumber(), b.GetSender(), level)
+		case constants.ApproveOnce:
+			// determine if build sender is in the contributors list for the repo
+			//
+			// NOTE: this call is cumbersome for repos with lots of contributors. Potential TODO: improve this if
+			// GitHub adds a single-contributor API endpoint.
+			contributor, err := scm.FromContext(c).RepoContributor(c.Request.Context(), r.GetOwner(), b.GetSender(), r.GetOrg(), r.GetName())
+			if err != nil {
+				return err
+			}
+
+			if !contributor {
+				err = gatekeepBuild(c, b, r)
+				if err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			fallthrough
+		case constants.ApproveNever:
+			fallthrough
+		default:
+			l.Debugf("fork PR build %s/%d automatically running without approval", r.GetFullName(), b.GetNumber())
+		}
+	}
+
+	// send API call to set the status on the commit
+	err := scm.FromContext(c).Status(c.Request.Context(), r.GetOwner(), b, r.GetOrg(), r.GetName())
+	if err != nil {
+		l.Errorf("unable to set commit status for %s/%d: %v", r.GetFullName(), b.GetNumber(), err)
+	}
+
+	// publish the build to the queue
+	go Enqueue(
+		context.WithoutCancel(c.Request.Context()),
+		queue.FromGinContext(c),
+		database.FromContext(c),
+		item,
+		b.GetHost(),
+	)
```

Comment:
thoughts on separating the logic from the action taken? 
> 	// determine whether to send compiled build to queue 

do the determining, `return bool error` then gatekeep or enqueue afterwards?
seems cleaner that way, like how we separate logic and API response writer code in other spots

Reply:
returning bool based on some logic might also make the function more testable

Reply:
I like the idea. There's a few spots within TrafficBuild and gatekeepBuild that leverage contextual services, which may make testing a little tricky. But I'll see what I can do.